### DISCUSSION TITLE
Content api content/v4/read is failing at local instance

### DISCUSSION THREAD

**Discussion by arun1205** [Sunbird-Knowlg/Community#118] [Bugs]
I am not able to load pdf content and video content to the player as the api that gets these contents is failing at local instance
I am pointing to dockstaging and  using all the env variables for the same
The entire url is as follows : http://localhost:3000/action/content/v4/read/do_213759703170523136117418

At dockstaging portal , its working https://dockstaging.sunbirded.org/action/content/v4/read/do_213759703170523136117418

@vinukumar-vs @pallakartheekreddy

**Comment by pallakartheekreddy** [comment]
@arun1205 Can you share the error log what you are getting? and are running Knowlg content-service in your local?

**Reply by arun1205** [reply]
@pallakartheekreddy  Its a 404 error that i am getting. I am not running Knowlg content-service  locally, but pointing to dockstaging.sunbirded.org
![image](https://user-images.githubusercontent.com/92018797/228740481-b6cd8251-4495-434f-b0f6-8dcda38db88e.png)

**Reply by arun1205** [reply]
@pallakartheekreddy @shubhambansaltarento

### Request flow from cokreat app to Knowledge middleware’s content service

Content Service is a part of Knowledge middleware services
Requests from the co-kreat application reaches the knowledge middleware via two ways
Viz
Case 1 : via Kong API gateway
Case 2 :  Directly to Knowledge Middleware 

### The request flow is as follows in Case 1

Cokreat client Application    ->    Cokreat  Node proxy server          ->      Kong API gateway     ->   Knowledge Middleware

When the suffix /`api` is added after the domain, the request follows the path mentioned in case 1
When the Kong api gateway receives this request and if this request contain v2 in the base url ,
only then it would redirect this request to knowledge middleware, else a 404 error would be thrown

For eg :`  dockstaging.sunbirded.org/api/content/v4/read` will throw an error as its taking the path mentioned in case 1.There is no configuration mapped in the kong for requests containing v4 in kong api gateway and hence 404

`dockstaging.sunbirded.org/api/content/v2/read` will return valid data as in the kong api gateway we can see the mapping for v2 which redirects the request to v4 of the content service

### The request flow is as follows in Case 2

Cokreat client Application       ->    Cokreat  Node proxy server         ->   Knowledge Middleware

When the suffix /`api` is NOT ADDED after the domain, the request follows the path mentioned in case 2
Since this request is bypassing the Kong api gateway and hitting the knowledge middleware directly as server to server communication and if this request contain  v4 in the base url ,
only then it would reach to knowledge middleware, else a 404 error would be thrown

The url mapping in Kong gateway is as follows
https://github.com/project-sunbird/sunbird-devops/blob/release-5.2.0/ansible/roles/kong-api/defaults/main.yml#L2681

https://github.com/project-sunbird/sunbird-devops/blob/release-5.2.0/ansible/roles/kong-api/defaults/main.yml#L9161



### /action conundrum in node proxy server (only for dev setup)

All the requests which contain /`action` from client app, re-routes via the node proxy server
This is the proxy server which is intended to obfuscate the actual url from the end user by adding /action

This routing is further explained in `contentEditorProxy.js`

**Reply by pallakartheekreddy** [reply]
@arun1205 is issue fixed? Or still facing issue while content v4 rad?

**Reply by arun1205** [reply]
@pallakartheekreddy after implementing these changes, its working as expected
